File: bestia-models/src/test/java/net/bestia/model/domain/GeneralModelTest.java
package net.bestia.model.domain;

import java.io.Serializable;
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

import javax.persistence.Entity;

import net.bestia.model.account.LoginInfo;
import org.junit.Assert;
import org.junit.Test;
import org.reflections.Reflections;

import com.fasterxml.jackson.annotation.JsonInclude.Include;
import com.fasterxml.jackson.databind.ObjectMapper;

public class GeneralModelTest {

	private ObjectMapper mapper = new ObjectMapper();
	private Reflections reflections = new Reflections("net.bestia.model");

	private final static Set<Class<? extends Object>> WHITELIST = new HashSet<>(Arrays.asList(LoginInfo.class));

	private Set<Class<?>> getAllEntities() {
		final Set<Class<?>> allClasses = reflections
				.getTypesAnnotatedWith(Entity.class);
		return allClasses;
	}
	
	public GeneralModelTest() {
		
		// Ignore null fields for the sake of the tests.
		mapper.setSerializationInclusion(Include.NON_NULL);
	}

	/**
	 * All entities must implement serializable.
	 */
	@Test
	public void all_serializable() {
		for (Class<?> clazz : getAllEntities()) {

			// Whitelist classes dont need to be serializable.
			if (WHITELIST.contains(clazz)) {
				continue;
			}

			Assert.assertTrue(clazz.toGenericString()
					+ " does not implement Serializable.",
					Serializable.class.isAssignableFrom(clazz));
		}
	}

	/**
	 * All entities must implement serializable.
	 * 
	 * @throws IllegalAccessException
	 * @throws InstantiationException
	 */
	@Test
	public void all_std_ctor() throws InstantiationException, IllegalAccessException {
		for (Class<?> clazz : getAllEntities()) {
			clazz.newInstance();
		}
	}

	/**
	 * All models should be serializable to JSON.
	 */
	@Test
	public void all_serializable_json() throws Exception {
		for (Class<?> clazz : getAllEntities()) {

			// Whitelist classes dont need to be serializable.
			if (WHITELIST.contains(clazz)) {
				continue;
			}

			Object obj = clazz.newInstance();
			mapper.writeValueAsString(obj);
		}
	}
}

File: bestia-models/src/test/kotlin/net/bestia/model/GeneralModelTest.kt
package net.bestia.model

import com.fasterxml.jackson.annotation.JsonInclude.Include
import com.fasterxml.jackson.databind.ObjectMapper
import net.bestia.model.bestia.PlayerBestia
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import org.reflections.Reflections
import java.io.Serializable
import javax.persistence.Entity

class GeneralModelTest {
  private val mapper = ObjectMapper().apply {
    // Ignore null fields for the sake of the tests.
    setSerializationInclusion(Include.NON_NULL)
  }

  private val reflections = Reflections("net.bestia.model")
  private val allEntities = reflections.getTypesAnnotatedWith(Entity::class.java)


  /**
   * All entities must implement serializable.
   */
  @Test
  fun all_serializable() {
    for (clazz in allEntities) {

      // Whitelist classes dont need to be serializable.
      if (WHITELIST.contains(clazz)) {
        continue
      }

      assertTrue(
          Serializable::class.java.isAssignableFrom(clazz),
          clazz.toGenericString() + " does not implement Serializable."
      )
    }
  }

  /**
   * All entities must implement serializable
   */
  @Test
  @Throws(InstantiationException::class, IllegalAccessException::class)
  fun all_std_ctor() {
    for (clazz in allEntities) {
      clazz.getDeclaredConstructor().newInstance()
    }
  }

  /**
   * All models should be serializable to JSON.
   */
  @Test
  @Throws(Exception::class)
  fun all_serializable_json() {
    for (clazz in allEntities) {

      // Whitelist classes dont need to be serializable.
      if (WHITELIST.contains(clazz)) {
        continue
      }

      val obj = clazz.getDeclaredConstructor().newInstance()
      mapper.writeValueAsString(obj)
    }
  }

  companion object {
    private val WHITELIST = setOf(
        PlayerBestia::class.java
    )
  }
}
